File: typescript/src/bittorrent/dht.ts
// SPDX-License-Identifier: MIT
// DHT (BEP-5): XOR distance + compact node (26B) / peer (6B) info.

export interface DhtContact {
  id: Buffer;
  ip: string;
  port: number;
}

export interface PeerAddr {
  ip: string;
  port: number;
}

export function xorDistance(a: Buffer, b: Buffer): Buffer {
  const out = Buffer.alloc(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] ^ b[i];
  return out;
}

function ipBytes(ip: string): Buffer {
  return Buffer.from(ip.split('.').map((x) => parseInt(x, 10)));
}

function ipStr(b: Buffer): string {
  return `${b[0]}.${b[1]}.${b[2]}.${b[3]}`;
}

export function encodeCompactNodes(nodes: DhtContact[]): Buffer {
  const parts: Buffer[] = [];
  for (const c of nodes) {
    const p = Buffer.alloc(2);
    p.writeUInt16BE(c.port, 0);
    parts.push(c.id, ipBytes(c.ip), p);
  }
  return Buffer.concat(parts);
}

export function decodeCompactNodes(data: Buffer): DhtContact[] {
  if (data.length % 26 !== 0) throw new Error('compact nodes length is not a multiple of 26');
  const out: DhtContact[] = [];
  for (let i = 0; i < data.length; i += 26) {
    out.push({
      id: Buffer.from(data.subarray(i, i + 20)),
      ip: ipStr(data.subarray(i + 20, i + 24)),
      port: data.readUInt16BE(i + 24),
    });
  }
  return out;
}

export function encodeCompactPeers(peers: PeerAddr[]): Buffer {
  const parts: Buffer[] = [];
  for (const p of peers) {
    const port = Buffer.alloc(2);
    port.writeUInt16BE(p.port, 0);
    parts.push(ipBytes(p.ip), port);
  }
  return Buffer.concat(parts);
}

export function decodeCompactPeers(data: Buffer): PeerAddr[] {
  if (data.length % 6 !== 0) throw new Error('compact peers length is not a multiple of 6');
  const out: PeerAddr[] = [];
  for (let i = 0; i < data.length; i += 6) {
    out.push({ ip: ipStr(data.subarray(i, i + 4)), port: data.readUInt16BE(i + 4) });
  }
  return out;
}

File: rust/src/bittorrent/dht.rs
// SPDX-License-Identifier: MIT
//! DHT (BEP-5): XOR distance + compact node (26B) / peer (6B) info.

pub fn xor_distance(a: &[u8], b: &[u8]) -> Vec<u8> {
    a.iter().zip(b).map(|(x, y)| x ^ y).collect()
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DhtContact {
    pub id: Vec<u8>,
    pub ip: String,
    pub port: u16,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PeerAddr {
    pub ip: String,
    pub port: u16,
}

fn ip_bytes(ip: &str) -> Vec<u8> {
    ip.split('.').map(|x| x.parse::<u8>().unwrap_or(0)).collect()
}

fn ip_str(b: &[u8]) -> String {
    format!("{}.{}.{}.{}", b[0], b[1], b[2], b[3])
}

pub fn encode_compact_nodes(nodes: &[DhtContact]) -> Vec<u8> {
    let mut out = Vec::with_capacity(nodes.len() * 26);
    for c in nodes {
        out.extend_from_slice(&c.id);
        out.extend_from_slice(&ip_bytes(&c.ip));
        out.extend_from_slice(&c.port.to_be_bytes());
    }
    out
}

pub fn decode_compact_nodes(data: &[u8]) -> Result<Vec<DhtContact>, String> {
    if data.len() % 26 != 0 {
        return Err("compact nodes length is not a multiple of 26".into());
    }
    Ok(data
        .chunks(26)
        .map(|c| DhtContact {
            id: c[0..20].to_vec(),
            ip: ip_str(&c[20..24]),
            port: u16::from_be_bytes([c[24], c[25]]),
        })
        .collect())
}

pub fn encode_compact_peers(peers: &[PeerAddr]) -> Vec<u8> {
    let mut out = Vec::with_capacity(peers.len() * 6);
    for p in peers {
        out.extend_from_slice(&ip_bytes(&p.ip));
        out.extend_from_slice(&p.port.to_be_bytes());
    }
    out
}

pub fn decode_compact_peers(data: &[u8]) -> Result<Vec<PeerAddr>, String> {
    if data.len() % 6 != 0 {
        return Err("compact peers length is not a multiple of 6".into());
    }
    Ok(data
        .chunks(6)
        .map(|c| PeerAddr {
            ip: ip_str(&c[0..4]),
            port: u16::from_be_bytes([c[4], c[5]]),
        })
        .collect())
}
